### PR TITLE
Expose query accesses

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

It would be useful to be able to inspect a `QueryState`'s accesses so we can detect when the data it accesses changes without having to iterate it. However there are two things preventing this:

* These accesses are unnecessarily encapsulated.
* `Has<T>` indirectly accesses `T`, but does not register it.

## Solution

* Expose accesses and matches used by `QueryState`.
* Add the notion of "archetypal" accesses, which are not accessed directly, but whose presence in an archetype affects a query result.

### CODE REVIEW

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:122]:
```suggestion
    /// Adds an archetypal (indirect) access to the element given by `index`.
```

Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:122]: These docs should capture the same information as `has_archetypal`

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:295]: Explanatory docs should be synced up here too.

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:498]: Naming and docs could be clearer here and below.

File: crates/bevy_ecs/src/query/state.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:99]:
```suggestion
    /// Returns the archetype-components accessed by this query.
```

**Review by alice-i-cecile — COMMENTED** [review]
And so we inch closer to the access model laid out in https://github.com/bevyengine/rfcs/pull/30 :D

I agree with this direction, and this PR is useful both as an ECS primitive and as something we can expose to users. I'd like to clean up the explanation and docs a bit further before merging.

Ideally we'd have module level documentation explaining the different kinds of access and how they relate to each other, but I won't block on that in this PR.

File: crates/bevy_ecs/src/query/state.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/state.rs:99]: Hmm, "archetype component" is a compound noun which I feel usually reads better without a hyphen. I prefer to save hyphens for compound adjectives

File: crates/bevy_ecs/src/query/access.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/query/access.rs]: Let me know if it's still not clear enough

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:124]:
```suggestion
    /// This is for elements whose values are not accessed (and thus will never cause conflicts),
```

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:153]:
```suggestion
    /// This is an element whose value is not accessed (and thus will never cause conflicts),
```

File: crates/bevy_ecs/src/query/access.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:298]:
```suggestion
    /// These are elements whose values are not accessed (and thus will never cause conflicts),
```

**Review by alice-i-cecile — APPROVED** [review]
Solid stuff. One small set of suggestions to clarify further, but I'm happy with this now.

**Comment by JoJoJet** [comment]
> Ideally we'd have module level documentation explaining the different kinds of access and how they relate to each other, but I won't block on that in this PR.

I'll think on what I would write here, but I probably won't include anything in this PR since module-level documentation generally requires a lot more thought than method-level documentation.

File: crates/bevy_ecs/src/query/access.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/access.rs:63]: Shouldn't these be populated by the appropriate QueryData/QueryFilter/SystemParam implementors? or is the plan to handle that in a separate PR?

Comment (JoJoJet) [reply crates/bevy_ecs/src/query/access.rs:63]: Oh right, forgot to move that over from my branch.

**Review by james7132 — APPROVED** [review]

**Pull request merged**
